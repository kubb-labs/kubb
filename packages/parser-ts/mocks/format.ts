import { format as prettierFormat } from 'prettier'
import pluginTypescript from 'prettier/plugins/typescript'

import type { Options } from 'prettier'

const formatOptions: Options = {
  tabWidth: 2,
  printWidth: 160,
  parser: 'typescript',
  singleQuote: true,
  semi: false,
  bracketSameLine: false,
  endOfLine: 'auto',
  plugins: [pluginTypescript],
}
export function format(source?: string): Promise<string> {
  if (!source) {
    return Promise.resolve('')
  }
  return prettierFormat(source, formatOptions)
}
