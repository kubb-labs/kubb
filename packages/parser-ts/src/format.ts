import { format as prettierFormat } from 'prettier'
import pluginTypescript from 'prettier/plugins/typescript'

import type { Options } from 'prettier'

const formatOptions: Options = {
  tabWidth: 2,
  printWidth: 160,
  parser: 'typescript',
  singleQuote: true,
  semi: false,
  bracketSameLine: false,
  endOfLine: 'auto',
  plugins: [pluginTypescript],
}

/**
 * Format the generated code based on Prettier
 */
export async function format(source: string) {
  // do some basic linting with the ts compiler
  if (!source) {
    return ''
  }

  return prettierFormat(source, formatOptions)
}
