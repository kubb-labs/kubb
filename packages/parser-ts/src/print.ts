import ts, { LanguageVariant } from 'typescript'

import type { PrinterOptions } from 'typescript'

const { factory } = ts

export type PrintOptions = {
  source?: string
  baseName?: string
  scriptKind?: ts.ScriptKind
} & PrinterOptions

/**
 * Escaped new lines in code with block comments so they can be restored by {@link restoreNewLines}
 * @param {string} code The code to escape new lines in
 * @returns The same code but with new lines escaped using block comments
 */
const escapeNewLines = (code: string) => code.replace(/\n\n/g, '\n/* :newline: */')

/**
 * Reverses {@link escapeNewLines} and restores new lines
 * @param {string} code The code with escaped new lines
 * @returns The same code with new lines restored
 */
const restoreNewLines = (code: string) => code.replace(/\/\* :newline: \*\//g, '\n')

/**
 * Convert AST TypeScript nodes to a string based on the TypeScript printer.
 */
export function print(
  elements: Array<ts.Node>,
  { source = '', baseName = 'print.ts', removeComments, noEmitHelpers, newLine = ts.NewLineKind.LineFeed, scriptKind = ts.ScriptKind.TS }: PrintOptions = {},
): string {
  const sourceFile = ts.createSourceFile(baseName, escapeNewLines(source), ts.ScriptTarget.ES2022, false, scriptKind)
  const printer = ts.createPrinter({
    omitTrailingSemicolon: true,
    newLine,
    removeComments,
    noEmitHelpers,
  })

  let nodes: Array<ts.Node> = []

  if (!elements) {
    return ''
  }

  if (Array.isArray(elements)) {
    nodes = elements.filter(Boolean)
  } else {
    nodes = [elements].filter(Boolean)
  }

  const outputFile = printer.printList(ts.ListFormat.MultiLine, factory.createNodeArray(nodes), sourceFile)
  const outputSource = printer.printFile(sourceFile)

  return [outputFile, restoreNewLines(outputSource)].filter(Boolean).join('\n')
}
