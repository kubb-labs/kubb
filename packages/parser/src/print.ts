import ts from 'typescript'

const { factory } = ts

export function print(elements: ts.Node | Array<ts.Node | undefined> | null, source = '', baseName = 'print.ts'): string {
  const printer = ts.createPrinter({ omitTrailingSemicolon: false, newLine: ts.NewLineKind.LineFeed })
  const sourceFile = ts.createSourceFile(baseName, source, ts.ScriptTarget.ESNext, true, ts.ScriptKind.TS)

  let nodes: Array<ts.Node> = []

  if (!elements) {
    return ''
  }

  if (Array.isArray(elements)) {
    nodes = elements.filter(Boolean) as ts.Node[]
  } else {
    nodes = [elements].filter(Boolean)
  }

  const outputFile = printer.printList(ts.ListFormat.MultiLine, factory.createNodeArray(nodes), sourceFile)
  const outputSource = printer.printFile(sourceFile)

  return [outputFile, outputSource].filter(Boolean).join('\n')
}
