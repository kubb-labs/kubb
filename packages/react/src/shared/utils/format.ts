import { print } from '@kubb/parser'

import type { Language } from '../../components/Editor.tsx'
import type { DOMElement } from '../../types.ts'

export function format(text: string, node: DOMElement): string {
  const attributes = node.attributes as React.ComponentProps<(typeof Language)>
  try {
    if (attributes.value === 'typescript') {
      return print([], { source: text, noEmitHelpers: false })
    }

    return text
  } catch (e) {
    console.log(e)
  }

  return text
}
