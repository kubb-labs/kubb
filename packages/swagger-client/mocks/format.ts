import { format as prettierFormat } from 'prettier'
import pluginTypescript from 'prettier/plugins/typescript'

import type { Options } from 'prettier'

const formatOptions: Options = {
  tabWidth: 2,
  printWidth: 160,
  parser: 'typescript',
  singleQuote: true,
  semi: false,
  bracketSameLine: false,
  endOfLine: 'auto',
  plugins: [pluginTypescript],
}
export function format(source?: string) {
  if (!source) {
    return ''
  }
  return prettierFormat(source, formatOptions)
}
