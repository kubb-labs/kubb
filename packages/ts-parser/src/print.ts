import ts from 'typescript'

import type { PrinterOptions } from 'typescript'

const { factory } = ts

type Options = {
  source?: string
  baseName?: string
} & PrinterOptions

/**
 * Escaped new lines in code with block comments so they can be restored by {@link restoreNewLines}
 * @param {string} code The code to escape new lines in
 * @returns The same code but with new lines escaped using block comments
 */
const escapeNewLines = (code: string) => code.replace(/\n\n/g, '\n/* :newline: */')

/**
 * Reverses {@link escapeNewLines} and restores new lines
 * @param {string} code The code with escaped new lines
 * @returns The same code with new lines restored
 */
const restoreNewLines = (code: string) => code.replace(/\/\* :newline: \*\//g, '\n')

export function print(
  elements: ts.Node | Array<ts.Node | undefined> | null,
  { source = '', baseName = 'print.ts', removeComments, noEmitHelpers, newLine = ts.NewLineKind.LineFeed }: Options = {},
): string {
  const sourceFile = ts.createSourceFile(baseName, escapeNewLines(source), ts.ScriptTarget.ES2022, false, ts.ScriptKind.TS)
  const printer = ts.createPrinter({
    omitTrailingSemicolon: true,
    newLine,
    removeComments,
    noEmitHelpers,
  })

  let nodes: Array<ts.Node> = []

  if (!elements) {
    return ''
  }

  if (Array.isArray(elements)) {
    nodes = elements.filter(Boolean)
  } else {
    nodes = [elements].filter(Boolean)
  }

  const outputFile = printer.printList(ts.ListFormat.MultiLine, factory.createNodeArray(nodes), sourceFile)
  const outputSource = printer.printFile(sourceFile)

  return [outputFile, restoreNewLines(outputSource)].filter(Boolean).join('\n')
}
