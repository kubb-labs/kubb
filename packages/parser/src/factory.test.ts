import ts from 'typescript'

import { format as prettierFormat } from '../mocks/format.ts'
import {
  appendJSDocToNode,
  createArrayDeclaration,
  createEnumDeclaration,
  createExportDeclaration,
  createImportDeclaration,
  createIntersectionDeclaration,
  createJSDoc,
  createParameterSignature,
  createPropertySignature,
  createQuestionToken,
  createUnionDeclaration,
  modifiers,
} from './factory.ts'
import { print } from './print.ts'

const { factory } = ts

describe('codegen', () => {
  const node = factory.createVariableStatement(
    undefined,
    factory.createVariableDeclarationList(
      [factory.createVariableDeclaration(factory.createIdentifier('hello'), undefined, undefined, factory.createStringLiteral('world'))],
      ts.NodeFlags.Const,
    ),
  )

  const formatTS = (elements: ts.Node | (ts.Node | undefined)[]) => {
    return prettierFormat(print(elements))
  }

  test('createQuestionToken', () => {
    expect(createQuestionToken()).toBeUndefined()
    expect(createQuestionToken(true)).toBeDefined()
  })

  test('createArrayDeclaration', () => {
    expect(
      print(
        createArrayDeclaration({
          nodes: [factory.createKeywordTypeNode(ts.SyntaxKind.StringKeyword), factory.createKeywordTypeNode(ts.SyntaxKind.NumberKeyword)],
        }),
      ),
    ).toMatchSnapshot()

    expect(
      print(
        createArrayDeclaration({
          nodes: [factory.createKeywordTypeNode(ts.SyntaxKind.StringKeyword)],
        }),
      ),
    ).toMatchSnapshot()
  })

  test('createIntersectionDeclaration', () => {
    expect(
      print(
        createIntersectionDeclaration({
          nodes: [factory.createKeywordTypeNode(ts.SyntaxKind.StringKeyword), factory.createKeywordTypeNode(ts.SyntaxKind.NumberKeyword)],
        }),
      ),
    ).toMatchSnapshot()
  })
  test('createUnionDeclaration', () => {
    expect(
      print(
        createUnionDeclaration({
          nodes: [factory.createKeywordTypeNode(ts.SyntaxKind.StringKeyword), factory.createKeywordTypeNode(ts.SyntaxKind.NumberKeyword)],
        }),
      ),
    ).toMatchSnapshot()
  })
  test('createPropertySignature', () => {
    expect(
      print(
        createPropertySignature({
          modifiers: [modifiers.const],
          name: 'hello',
          type: factory.createKeywordTypeNode(ts.SyntaxKind.StringKeyword),
        }),
      ),
    ).toMatchSnapshot()
  })

  test('createParameter', () => {
    expect(
      print(
        createParameterSignature('hello', {
          type: factory.createKeywordTypeNode(ts.SyntaxKind.StringKeyword),
        }),
      ),
    ).toMatchSnapshot()
    expect(
      print(
        createParameterSignature('hello', {
          questionToken: true,
          type: factory.createKeywordTypeNode(ts.SyntaxKind.StringKeyword),
        }),
      ),
    ).toMatchSnapshot()
    expect(
      print(
        createParameterSignature('hello', {
          questionToken: true,
          type: factory.createKeywordTypeNode(ts.SyntaxKind.BooleanKeyword),
        }),
      ),
    ).toMatchSnapshot()
  })

  test('createJSDoc', async () => {
    expect(
      await formatTS(
        createJSDoc({
          comments: ['@description description', '@example example'],
        })!,
      ),
    ).toMatchSnapshot()
  })

  test('appendJSDocToNode', async () => {
    expect(
      await formatTS(
        appendJSDocToNode({
          node: { ...node },
          comments: ['@description description', undefined, '@example example'],
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        appendJSDocToNode({
          node: { ...node },
          comments: [],
        }),
      ),
    ).toMatchSnapshot()
  })

  test('createImportDeclaration', async () => {
    expect(
      await formatTS(
        createImportDeclaration({
          name: 'hello',
          path: './hello.ts',
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createImportDeclaration({
          name: 'hello',
          path: './hello.ts',
          isTypeOnly: true,
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createImportDeclaration({
          name: ['hello'],
          path: './hello.ts',
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createImportDeclaration({
          name: 'hello',
          path: './hello.ts',
          isNameSpace: true,
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createImportDeclaration({
          name: [{ propertyName: 'hello', name: 'helloWorld' }],
          path: './hello.ts',
        }),
      ),
    ).toMatchSnapshot()
  })

  test('createExportDeclaration', async () => {
    expect(
      await formatTS(
        createExportDeclaration({
          path: './hello.ts',
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createExportDeclaration({
          name: ['hello', 'world'],
          asAlias: true,
          path: './hello.ts',
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createExportDeclaration({
          name: 'hello',
          asAlias: true,
          path: './hello.ts',
        }),
      ),
    ).toMatchSnapshot()

    try {
      await formatTS(
        createExportDeclaration({
          name: 'hello',
          path: './hello.ts',
        }),
      )
    } catch (e) {
      expect(e).toBeDefined()
    }
  })

  test('createEnumDeclaration', async () => {
    expect(
      await formatTS(
        createEnumDeclaration({
          type: 'enum',
          name: 'hello',
          typeName: 'Hello',
          enums: [
            ['hello', 'world'],
            ['end', 2050],
            ['survive', true],
          ],
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createEnumDeclaration({
          type: 'asConst',
          name: 'hello',
          typeName: 'Hello',
          enums: [
            ['hello', 'world'],
            ['end', 2050],
            ['survive', true],
          ],
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createEnumDeclaration({
          type: 'asConst',
          name: 'hello',
          typeName: 'Hello',
          enums: [
            ['FILE.UPLOADED', 'FILE.UPLOADED'],
            ['FILE.PREVIEWE', 'FILE.PREVIEWE'],
          ],
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createEnumDeclaration({
          type: 'asPascalConst',
          name: 'hello',
          typeName: 'Hello',
          enums: [
            ['hello', 'world'],
            ['end', 2050],
            ['survive', true],
          ],
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createEnumDeclaration({
          type: 'constEnum',
          name: 'hello',
          typeName: 'Hello',
          enums: [
            ['hello', 'world'],
            ['end', 2050],
            ['survive', true],
          ],
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createEnumDeclaration({
          type: 'literal',
          name: 'hello',
          typeName: 'Hello',
          enums: [
            ['hello', 'world'],
            ['end', 2050],
            ['survive', true],
          ],
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createEnumDeclaration({
          type: 'enum',
          name: 'hello',
          typeName: 'Hello',
          enums: [
            ['1', 'world'],
            ['2', 2050],
            ['3', true],
          ],
        }),
      ),
    ).toMatchSnapshot()
  })
})
