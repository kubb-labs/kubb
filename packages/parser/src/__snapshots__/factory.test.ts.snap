// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`codegen > appendJSDocToNode 1`] = `
"/**
 * @description description
 * @example example
 */
const hello = 'world'
"
`;

exports[`codegen > appendJSDocToNode 2`] = `
"const hello = 'world'
"
`;

exports[`codegen > createArrayDeclaration 1`] = `
"Array<string | number>
"
`;

exports[`codegen > createArrayDeclaration 2`] = `
"string[]
"
`;

exports[`codegen > createEnumDeclaration 1`] = `
"export enum Hello {
  'hello' = 'world',
  'end' = 2050,
  'survive' = true,
}
"
`;

exports[`codegen > createEnumDeclaration 2`] = `
"export const hello = {
  hello: 'world',
  end: 2050,
  survive: true,
} as const
export type Hello = (typeof hello)[keyof typeof hello]
"
`;

exports[`codegen > createEnumDeclaration 3`] = `
"export const hello = {
  'FILE.UPLOADED': 'FILE.UPLOADED',
  'FILE.PREVIEWE': 'FILE.PREVIEWE',
} as const
export type Hello = (typeof hello)[keyof typeof hello]
"
`;

exports[`codegen > createEnumDeclaration 4`] = `
"export const Hello = {
  hello: 'world',
  end: 2050,
  survive: true,
} as const
export type Hello = (typeof Hello)[keyof typeof Hello]
"
`;

exports[`codegen > createEnumDeclaration 5`] = `
"export const enum Hello {
  'hello' = 'world',
  'end' = 2050,
  'survive' = true,
}
"
`;

exports[`codegen > createEnumDeclaration 6`] = `
"export type Hello = 'world' | 2050 | true
"
`;

exports[`codegen > createEnumDeclaration 7`] = `
"export enum Hello {
  'Hello_1' = 'world',
  'Hello_2' = 2050,
  'Hello_3' = true,
}
"
`;

exports[`codegen > createExportDeclaration 1`] = `
"export * from './hello.ts'
"
`;

exports[`codegen > createExportDeclaration 2`] = `
"export { hello, world } from './hello.ts'
"
`;

exports[`codegen > createExportDeclaration 3`] = `
"export * as hello from './hello.ts'
"
`;

exports[`codegen > createImportDeclaration 1`] = `
"import hello from './hello.ts'
"
`;

exports[`codegen > createImportDeclaration 2`] = `
"import type hello from './hello.ts'
"
`;

exports[`codegen > createImportDeclaration 3`] = `
"import { hello } from './hello.ts'
"
`;

exports[`codegen > createImportDeclaration 4`] = `
"import * as hello from './hello.ts'
"
`;

exports[`codegen > createImportDeclaration 5`] = `
"import { hello as helloWorld } from './hello.ts'
"
`;

exports[`codegen > createIntersectionDeclaration 1`] = `
"string & number
"
`;

exports[`codegen > createJSDoc 1`] = `
"/**
 * @description description
 * @example example */
"
`;

exports[`codegen > createParameter 1`] = `
"hello: string
"
`;

exports[`codegen > createParameter 2`] = `
"hello?: string
"
`;

exports[`codegen > createParameter 3`] = `
"hello?: boolean
"
`;

exports[`codegen > createPropertySignature 1`] = `
"const hello: string;
"
`;

exports[`codegen > createUnionDeclaration 1`] = `
"string | number
"
`;
